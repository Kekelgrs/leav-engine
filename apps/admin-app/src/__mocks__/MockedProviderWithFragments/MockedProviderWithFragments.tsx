// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {MockedProvider} from '@apollo/react-testing';
import {InMemoryCache} from 'apollo-cache-inmemory';
import React from 'react';
import {attributesFragmentMatcher} from '../fragmentMatchers/attributesFragmentMatchers';

export const getMockCacheWithFragments = () => new InMemoryCache({fragmentMatcher: attributesFragmentMatcher});

function MockedProviderWithFragments({children, ...props}) {
    // Set a new cache for each test to avoid fetching data in cache and not in provided mocks
    const mockCache = getMockCacheWithFragments();
    return (
        <MockedProvider cache={mockCache} {...props}>
            {children}
        </MockedProvider>
    );
}

export default MockedProviderWithFragments;
