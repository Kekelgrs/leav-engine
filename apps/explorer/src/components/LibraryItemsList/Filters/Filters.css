.wrapped-filter-open {
    animation-name: filter-animation-open;
    animation-duration: 1s;
    animation-timing-function: ease;
}

@keyframes filter-animation-open {
    from {
        /* transform: translate(-50rem); */
        opacity: 0;
    }

    to {
        /* transform: translate(0); */
        opacity: 1;
    }
}

.wrapped-filter-close {
    animation-name: filter-animation-close;
    animation-duration: 1s;
    animation-timing-function: ease-out;
}

@keyframes filter-animation-close {
    from {
        /* transform: translate(0); */
        opacity: 1;
    }

    to {
        /* transform: translate(-50rem); */
        opacity: 0;
    }
}