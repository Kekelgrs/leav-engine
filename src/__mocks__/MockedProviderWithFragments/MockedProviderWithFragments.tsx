// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {InMemoryCache} from '@apollo/client';
import {MockedProvider} from '@apollo/client/testing';
import React from 'react';

// export const getMockCacheWithFragments = () => new InMemoryCache({fragmentMatcher: attributesFragmentMatcher});
export const getMockCacheWithFragments = () => new InMemoryCache();

function MockedProviderWithFragments({children, ...props}: any) {
    // Set a new cache for each test to avoid fetching data in cache and not in provided mocks
    const mockCache = getMockCacheWithFragments();
    return (
        <MockedProvider cache={mockCache} {...props}>
            {children}
        </MockedProvider>
    );
}

export default MockedProviderWithFragments;
