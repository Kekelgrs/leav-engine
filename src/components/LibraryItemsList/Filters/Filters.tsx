import {useQuery} from '@apollo/react-hooks';
import React, {useEffect, useState} from 'react';
import {
    Button,
    Container,
    Divider,
    Dropdown,
    Grid,
    Header,
    Input,
    List,
    Menu,
    Segment,
    Sidebar,
    Transition
} from 'semantic-ui-react';
import styled from 'styled-components';
import {getLibraryDetailQuery} from '../../../queries/libraries/getLibraryDetailQuery';
import SelectVue from '../SelectVue';

interface IFiltersProps {
    showFilters: boolean;
    setShowFilters: (showFilters: (x: boolean) => boolean) => void;
    libId: string;
    libQueryName: string;
}

const Side = styled.div`
    border: 1px solid #ebebeb;
    padding: 1rem 1rem 0 1rem;
    height: 100%;
`;

const whereOptions = [
    {text: 'Contain', value: 'contain'},
    {text: "Don't Contain", value: 'not-contain'},
    {text: 'Is Equal', value: 'is-equal'},
    {text: 'Begin With', value: 'begin-with'},
    {text: 'End With', value: 'end-with'},
    {text: 'Is Empty', value: 'empty'},
    {text: 'Is Not Empty', value: 'not-empty'},
    {text: 'Existing Value', value: 'exist'},
    {
        text: 'divider',
        value: 'divider',
        content: <Dropdown.Divider />,
        disabled: true
    },
    {text: 'Search In...', value: 'not-exist'}
];

interface IFilters {
    key: any;
    where: string;
    value: string;
    attribute: string;
}

function Filters({showFilters, setShowFilters, libId, libQueryName}: IFiltersProps): JSX.Element {
    const [showAttr, setShowAttr] = useState(false);
    const [attrs, setAttrs] = useState<any>([]);
    const [show, setShow] = useState(showFilters);
    const [filters, setFilters] = useState<IFilters[]>([
        {
            key: 1,
            where: 'contain',
            value: 'mon produit',
            attribute: 'ean'
        }
    ]);

    const {loading, data, error} = useQuery(getLibraryDetailQuery(libQueryName), {
        variables: {
            libId
        }
    });

    useEffect(() => {
        if (!loading) {
            const attributes = data?.libraries?.list[0]?.attributes;
            setAttrs(attributes);
        }
    }, [loading, data, error]);

    useEffect(() => {
        setShow(showFilters);
    }, [showFilters]);

    return (
        <Transition visible={show} onHide={() => setShowFilters(show => false)} animation="slide right" duration={100}>
            <Sidebar.Pushable>
                <Sidebar
                    visible={showAttr}
                    onHide={() => setShowAttr(false)}
                    animation="push"
                    width="wide"
                    direction="right"
                >
                    <Container>
                        <Button icon="close" onClick={() => setShowAttr(false)} />
                        <List>{attrs && attrs.map((att: any) => <List.Item key={att.id}>{att.id}</List.Item>)}</List>
                    </Container>
                </Sidebar>
                <Sidebar.Pusher>
                    <Side>
                        <Menu>
                            <Menu.Menu>
                                <Menu.Item>
                                    <Button icon="sidebar" onClick={() => setShow(false)} />
                                </Menu.Item>
                            </Menu.Menu>
                            <Menu.Menu position="right">
                                <Menu.Item>
                                    <SelectVue />
                                </Menu.Item>
                            </Menu.Menu>
                        </Menu>

                        <List horizontal>
                            <List.Item>
                                <Header as="h5">Filters</Header>
                            </List.Item>
                            <List.Item>
                                <Button basic icon="plus" onClick={() => setShowAttr(true)} />
                            </List.Item>
                        </List>

                        <Transition.Group>
                            <Segment secondary color="green">
                                <Grid columns={3}>
                                    <Grid.Row>
                                        <Grid.Column width="4">
                                            <strong>Where</strong>
                                        </Grid.Column>
                                        <Grid.Column width="4">
                                            <strong>Label</strong>
                                        </Grid.Column>
                                        <Grid.Column>
                                            <strong>Value</strong>
                                        </Grid.Column>
                                    </Grid.Row>
                                    {filters.map(filter => (
                                        <Grid.Row key={filter.key}>
                                            <Grid.Column width="4">
                                                <Dropdown
                                                    floating
                                                    inline
                                                    defaultValue={filter.where}
                                                    options={whereOptions}
                                                />
                                            </Grid.Column>
                                            <Grid.Column width="4">{filter.attribute}</Grid.Column>
                                            <Grid.Column width="8">
                                                <Input value={filter.value} fluid />
                                            </Grid.Column>
                                        </Grid.Row>
                                    ))}
                                </Grid>
                                <Divider />
                                <Grid columns={2}>
                                    <Grid.Column>
                                        <Button negative compact onClick={() => setFilters([])}>
                                            Remove filters
                                        </Button>
                                    </Grid.Column>
                                    <Grid.Column>
                                        <Button positive compact>
                                            Apply
                                        </Button>
                                    </Grid.Column>
                                </Grid>
                            </Segment>
                        </Transition.Group>
                    </Side>
                    <Transition visible={showAttr} animation="slide right" duration={100}>
                        <Segment>attr list </Segment>
                    </Transition>
                </Sidebar.Pusher>
            </Sidebar.Pushable>
        </Transition>
    );
}

export default Filters;
