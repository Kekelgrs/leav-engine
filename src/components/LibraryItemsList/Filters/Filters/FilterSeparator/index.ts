import FilterSeparator from './FilterSeparator';
export default FilterSeparator;
