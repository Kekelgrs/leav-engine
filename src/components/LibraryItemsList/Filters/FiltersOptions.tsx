import {TFunction} from 'i18next';
import {conditionFilter, operatorFilter} from '../../../_types/types';

export function getConditionOptions(t: TFunction) {
    const conditionOptions = [
        {text: t('filters.contains'), value: conditionFilter.contains},
        {text: t('filters.not-contains'), value: conditionFilter.notContains},
        {text: t('filters.equal'), value: conditionFilter.equal},
        {text: t('filters.not-equal'), value: conditionFilter.notEqual},
        {text: t('filters.begin-with'), value: conditionFilter.beginWith},
        {text: t('filters.end-with'), value: conditionFilter.endWith},
        {text: t('filters.is-empty'), value: conditionFilter.empty},
        {text: t('filters.is-not-empty'), value: conditionFilter.notEmpty},
        {text: t('filters.greater-than'), value: conditionFilter.greaterThan},
        {text: t('filters.less-than'), value: conditionFilter.lessThan},
        {text: t('filters.exist'), value: conditionFilter.exist},
        {text: t('filters.search-in'), value: conditionFilter.searchIn}
    ];

    return conditionOptions;
}

export function getOperatorOptions(t: TFunction) {
    const operatorOptions = [
        {text: t('filters.and'), value: operatorFilter.and},
        {text: t('filters.or'), value: operatorFilter.or}
    ];

    return operatorOptions;
}
