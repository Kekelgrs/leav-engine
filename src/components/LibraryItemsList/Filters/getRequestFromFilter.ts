import {FilterTypes, IFilter, IFilterSeparator, IQueryFilter, OperatorFilter} from '../../../_types/types';
import {AttributeFormat} from './../../../_types/types';

export const getRequestFromFilter = (
    filters: (IFilter | IFilterSeparator)[],
    filterOperator: OperatorFilter,
    separatorOperator: OperatorFilter
): IQueryFilter[] => {
    let request: IQueryFilter[] = [];

    const hasSeparator = filters.some(filter => filter.type === FilterTypes.separator);
    const firstFilterIsFilterWithValue = filters[0].type === FilterTypes.filter && filters[0].value;

    if (hasSeparator && firstFilterIsFilterWithValue) {
        // Begin first group of filters
        request.push({operator: OperatorFilter.openParent});
    }

    for (let filter of filters) {
        if (filter.type === FilterTypes.filter && filter.active && filter.value) {
            if (filter.operator) {
                request.push({operator: filterOperator});
            }
            request.push({operator: OperatorFilter.openParent});

            const requestValue = handleValueRequest(filter);
            request = [...request, ...requestValue];

            request.push({operator: OperatorFilter.closeParent});
        } else if (filter.type === FilterTypes.separator && firstFilterIsFilterWithValue) {
            if (
                filter.active &&
                (filters[filter.key + 1] as IFilter)?.active &&
                (filters[filter.key + 1] as IFilter)?.value
            ) {
                // Close last group
                request.push({operator: OperatorFilter.closeParent});

                request.push({operator: separatorOperator});

                // Begin new group
                request.push({operator: OperatorFilter.openParent});
            }
        }
    }
    if (hasSeparator && firstFilterIsFilterWithValue) {
        // Close last group
        request.push({operator: OperatorFilter.closeParent});
    }

    if (!request.some(el => el.value)) {
        request = [];
    }

    return request;
};

const handleValueRequest = (filter: IFilter) => {
    const result: IQueryFilter[] = [];
    let firstValue = true;

    switch (filter.format) {
        case AttributeFormat.text: {
            filter.value.split('\n').forEach(filterValue => {
                if (filterValue) {
                    if (!firstValue) {
                        result.push({operator: OperatorFilter.or});
                    }

                    result.push({
                        field: getValueFromFilter(filter),
                        value: filterValue,
                        condition: filter.condition
                    });
                    firstValue = false;
                }
            });
            break;
        }
        case AttributeFormat.boolean: {
            result.push({
                field: filter.attributeId,
                value: filter.value,
                condition: filter.condition
            });
            break;
        }
    }

    return result;
};

const getValueFromFilter = (filter: IFilter) => {
    if (filter.originAttributeData) {
        return `${filter.originAttributeData.id}.${filter.attributeId}`;
    }
    if (filter.extendedData) {
        return `${filter.extendedData.path}`;
    }
    return `${filter.attributeId}`;
};
