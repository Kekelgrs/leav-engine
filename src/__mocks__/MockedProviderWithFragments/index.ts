import MockedProviderWithFragments from './MockedProviderWithFragments';
export default MockedProviderWithFragments;
