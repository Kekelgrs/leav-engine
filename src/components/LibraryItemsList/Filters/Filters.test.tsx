import {render} from 'enzyme';
import React from 'react';
import MockedProviderWithFragments from '../../../__mocks__/MockedProviderWithFragments';
import Filters from './Filters';

describe('Filters', () => {
    test('Snapshot test', async () => {
        const comp = render(
            <MockedProviderWithFragments>
                <Filters showFilters={false} setShowFilters={jest.fn()} libId="test" libQueryName="test" />
            </MockedProviderWithFragments>
        );

        expect(comp).toMatchSnapshot();
    });
});
