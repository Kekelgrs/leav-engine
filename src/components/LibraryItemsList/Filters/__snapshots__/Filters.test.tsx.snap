// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters Snapshot test 1`] = `
Object {
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
